Add `dev_getBlockStats` RPC

athei opened the pull request:
This adds a new RPC:

```rust
pub struct BlockStats {
    pub witness_len: u64,
    pub witness_compact_len: u64,
    pub block_len: u64,
    pub block_num_extrinsics: u64,
}

#[rpc(name = "dev_getBlockStats")]
fn block_stats(&self, hash: Hash) -> Result<Option<BlockStats>>;
```

It is meant as a way to gain insides about the bottle necks of a given chain. While some information can only be learned through this RPC (witness related information) some of them are redundant: You can query the whole block and then measure the size of the block and the number of extrinsics. **However**, I think it still makes sense to include them as this saves us from transmitting the whole block to the RPC client (possibly over the internet) in order to learn about these statistics.

I have written a CLI tool that queries this draft version of the RPC:

```bash
cargo install --git https://github.com/athei/povstats.git --locked

# This will just connect to a locally running node and calls the RPC on each new block
povstats
```

The output when running this against a `substrate --dev` node without submitting any extrinsics to it looks like this:
```
27: Some(BlockStats { witness_len: 8179, witness_compact_len: 6124, witness_compressed_len: 5685, block_len: 226, block_num_extrinsics: 1 })
28: Some(BlockStats { witness_len: 7960, witness_compact_len: 5869, witness_compressed_len: 5412, block_len: 129, block_num_extrinsics: 1 })
29: Some(BlockStats { witness_len: 7894, witness_compact_len: 5901, witness_compressed_len: 5473, block_len: 129, block_num_extrinsics: 1 })
30: Some(BlockStats { witness_len: 8564, witness_compact_len: 6379, witness_compressed_len: 5918, block_len: 226, block_num_extrinsics: 1 })
31: Some(BlockStats { witness_len: 8535, witness_compact_len: 6382, witness_compressed_len: 5913, block_len: 226, block_num_extrinsics: 1 })
32: Some(BlockStats { witness_len: 8698, witness_compact_len: 6513, witness_compressed_len: 6065, block_len: 226, block_num_extrinsics: 1 })
33: Some(BlockStats { witness_len: 7842, witness_compact_len: 5913, witness_compressed_len: 5471, block_len: 129, block_num_extrinsics: 1 })
34: Some(BlockStats { witness_len: 7886, witness_compact_len: 5925, witness_compressed_len: 5473, block_len: 129, block_num_extrinsics: 1 })
35: Some(BlockStats { witness_len: 8520, witness_compact_len: 6497, witness_compressed_len: 6064, block_len: 226, block_num_extrinsics: 1 })
```

kianenigma commented:
> It happens inside: sp_api::Core::execute_block which is called by the new RPC.

I am not sure with the abstraction level that you are using, and I don't see any code that specifies on top of which state you are executing the given block. What I can say for sure is that to execute any arbitrary block, you certainly need its parent state, which may not be available. This error that you said also rings a bell for me in `try-runtime`, and I recall the reason was not using the proper parent state.

athei commented:
I thought I did correctly specify the parent state by calling the runtime API at the block's parent's hash:

```rust
runtime_api.execute_block(&BlockId::Hash(*parent_hash), block)
```

kianenigma commented:
> I thought I did correctly specify the parent state by calling the runtime API at the block's parent's hash:
> 
> ```rust
> runtime_api.execute_block(&BlockId::Hash(*parent_hash), block)
> ```

that sounds correct to me. Perhaps to double check: if you pass an old-block's hash to a non-archive node, do you get an error?

bkchr commented on client/rpc/src/chain/chain_full.rs:
```diff
@@ -69,4 +82,54 @@ where
 	}
+
+	fn block_stats(&self, hash: Option<Block::Hash>) -> Result<Option<BlockStats>> {
+		let block = {
```
Your problem is that the block header contains the seal of the consensus engine. The runtime is "dumb" it just compares that the digests of the header that is the result of `execute_block` matches the header of the block you gave as input. The seal is added by the consensus engine after the runtime and is normally stripped before we call into the runtime.

athei replied:
Is there anything I can do here? Should I just drop the last `Digest` using `block.header().digest_mut()`?

bkchr replied:
You could drop all of them that are `Digest::Seal`

athei replied:
Thanks a lot. I removed all of those before passing them into the runtime and it works now.

bkchr reviewed (changes requested):
In generally I'm against adding new rpc methods that are added for everybody by default. However, the functionality here could maybe added to the `state_traceBlock` rpc method. Or if you want to keep it separate, it should be put into its own crate.

athei commented:
> In generally I'm against adding new rpc methods that are added for everybody by default. However, the functionality here could maybe added to the state_traceBlock rpc method.

Isn't modifying existing RPCs that are included for anyone even worse? I can move it to its own crate so the RPC is optional.

bkchr commented:
> Isn't modifying existing RPCs that are included for anyone even worse?

I mean we would just add some data to the output that you can just ignore on the calling site.

athei commented:
> > Isn't modifying existing RPCs that are included for anyone even worse?
> 
> I mean we would just add some data to the output that you can just ignore on the calling site.

This is basically modifying the RPC in a backwards compatible way. So is adding a new RPC. Why is adding a new one worse?

dvdplm commented:
My two cents:
- We don't have a good process to decide when/how/who to add RPCs (same for deprecation).
- Over time I've seen people think that "unsafe" RPCs are a way to signal that an RPC is experimental/debug-only/unstable. This is not the case, or rather, it's a bit of a semantic acrobatics. Unsafe means that an RPC should not be exposed to untrusted input, either because it does dangerous things or because it puts a lot of strain on the node and thus exposing a DoS veector.
- @tomaka 's approach when re-designing the RPCs is to be very explicit about which RPCs are stable and which aren't, by including "unstable" in the name. I think that's one way to go about it (preferably letting the proc macro do the heavy lifting).
- Given that up until now, adding or changing an RPC has been pretty much "do whatever you want", I think it's unfair to block this PR and I don't think the functionality fits very well as an extension to `traceBlock`


In other words: no objections to merge this from me (but it'd be good to start making up our minds about how to add RPCs, the sooner the better).

bkchr commented:
> In other words: no objections to merge this from me (but it'd be good to start making up our minds about how to add RPCs, the sooner the better).

This is relative easy. Don't add RPCs automatically to every node building on Substrate. Adding RPC functionality to Substrate isn't bad and we can name them `unstable` or whatever. However, people building on Substrate should decide on their own what kind of RPCs they want to expose.

athei commented on client/rpc/src/chain/chain_full.rs:
```diff
@@ -69,4 +82,55 @@ where
 	}
+
+	fn block_stats(&self, hash: Option<Block::Hash>) -> Result<Option<BlockStats>> {
+		let block = {
```
Yes. PR is still WIP (draft). The unsafe check is missing.

athei replied:
Added the check.

athei commented on client/rpc/src/chain/mod.rs:
```diff
@@ -72,6 +75,8 @@ where
 	/// Get header and body of a relay chain block.
 	fn block(&self, hash: Option<Block::Hash>) -> FutureResult<Option<SignedBlock<Block>>>;
 
+	fn block_stats(&self, hash: Option<Block::Hash>) -> Result<Option<BlockStats>>;
```
Yes. I will add documentation once there is an agreement where to put this RPC.

athei commented:
> > In other words: no objections to merge this from me (but it'd be good to start making up our minds about how to add RPCs, the sooner the better).
> 
> This is relative easy. Don't add RPCs automatically to every node building on Substrate. Adding RPC functionality to Substrate isn't bad and we can name them `unstable` or whatever. However, people building on Substrate should decide on their own what kind of RPCs they want to expose.

So a new category within `sc-rpc` would be fine? Putting it into a new crate just adds more boilerplate for no benefit.

bkchr commented:
Yeah, a new category in `sc-rpc` is okay :) I mean, I have no idea how the structure with jsonrpsee looks, that was the reason I proposed to speak with @dvdplm :)

athei commented:
@dvdplm said: The less boilerplate the better for out transition. So even adding a new category has a lot of boilerplate and makes the transition more work. That said, we need to balance this against how bad it us for RPCs to "just show up" when you bump substrate.

dvdplm commented:
> I mean, I have no idea how the structure with jsonrpsee looks, that was the reason I proposed to speak with @dvdplm :)

The amount of work for devs is quite similar, maybe slightly less but not by much.

athei commented:
Okay so I will add a new category then. I still have no idea how to call this category. I was thinking about calling it `dev` to clearly indicate that it will contain some seriously wonky RPCs that are not really stable. We could then add new stuff there without much concern because wouldn't add this category to their production node anyways.

What do you think?

athei commented on primitives/runtime/src/lib.rs:
```diff
@@ -902,6 +902,17 @@ impl<R> TransactionOutcome<R> {
+	pub witness_compressed_len: u64,
+	pub block_len: u64,
+	pub block_num_extrinsics: u64,
+}
```
Added

athei commented:
Ready for review now:
* Moved RPC to a new category.
* Added docs.
* Removed the `witness_compressed_len`. It is too assumptions. Client can compress itself if wanted.

niklasad1 commented on client/rpc-api/src/dev/error.rs:
```diff
@@ -0,0 +1,71 @@
+		match e {
+			Error::BlockQueryError(_) => rpc::Error {
+				code: rpc::ErrorCode::ServerError(BASE_ERROR + 1),
+				message: format!("{}", e),
```
```suggestion
				message: e.to_string(),
```

niklasad1 commented on client/rpc-api/src/dev/error.rs:
```diff
@@ -0,0 +1,71 @@
+			},
+			Error::BlockExecutionFailed => rpc::Error {
+				code: rpc::ErrorCode::ServerError(BASE_ERROR + 3),
+				message: format!("{}", e),
```
```suggestion
				message: e.to_string(),
```

niklasad1 commented on client/rpc-api/src/dev/error.rs:
```diff
@@ -0,0 +1,71 @@
+			},
+			Error::WitnessCompactionFailed => rpc::Error {
+				code: rpc::ErrorCode::ServerError(BASE_ERROR + 4),
+				message: format!("{}", e),
```
```suggestion
				message: e.to_string(),
```

bkchr commented on primitives/runtime/src/lib.rs:
```diff
@@ -902,6 +902,27 @@ impl<R> TransactionOutcome<R> {
+	pub witness_len: u64,
+	/// The length in bytes of the storage proof after it was compacted.
+	pub witness_compact_len: u64,
+	/// Length of the block in bytes.
+	///
+	/// This information could also be acquired by downloading the whole block. This is merely
+	/// a convenience to keep the traffic between client and node low.
+	pub block_len: u64,
+	/// Number of extrinsics in the block.
+	///
+	/// This information could also be acquired by downloading the whole block. This is merely
+	/// a convenience to keep the traffic between client and node low.
+	pub num_extrinsics: u64,
```
For a dev api. Do we really need to keep the traffic low? Sounds like some "over optimization"

athei replied:
I would rather argue that leaving it out would be over optimization: It is very easy to include this data. If we'd leave it out the client needs to refetch the block which makes the RPC client more complicated and brittle.

The reason for leaving it out would be to deduplicate. Deduplication is just another form of optimization :)

athei replied:
I can update the comment.

bkchr commented on client/rpc/src/dev/mod.rs:
```diff
@@ -0,0 +1,116 @@
+				.map_err(|e| Error::BlockQueryError(Box::new(e)))?;
+			if let Some(block) = block {
+				let (mut header, body) = block.block.deconstruct();
+				header.digest_mut().logs.retain(|item| !matches!(item, DigestItem::Seal(_, _)));
```
```suggestion
				// Remove the `Seal` to ensure we have the number of digests as expected by the runtime.
				header.digest_mut().logs.retain(|item| !matches!(item, DigestItem::Seal(_, _)));
```

bkchr commented on primitives/runtime/src/lib.rs:
```diff
@@ -902,6 +902,27 @@ impl<R> TransactionOutcome<R> {
+#[derive(Eq, PartialEq, Clone, Copy, Encode, Decode, Debug, TypeInfo)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[cfg_attr(feature = "std", serde(rename_all = "camelCase"))]
+pub struct BlockStats {
```
And why is that defined in `sp-runtime`? Can that type not just be defined alongside the rpc api declaration?

athei replied:
I am totally lost where to define that. If you think it should be beside the RPC I can move it there. I just saw that no type was defined there and figured that this isn't the place.

bkchr replied:
This type doesn't make any sense outside of this rpc. So, we should define it near to the rpc api.

athei replied:
Moved it.

bkchr commented on client/rpc/src/dev/mod.rs:
```diff
@@ -0,0 +1,116 @@
+
+pub use sc_rpc_api::dev::DevApi;
+
+use sc_client_api::BlockBackend;
```
```suggestion
use sc_client_api::{BlockBackend, HeaderBackend};
```

bkchr commented on client/rpc/src/dev/mod.rs:
```diff
@@ -0,0 +1,116 @@
+impl<Block, Client> DevApi<Block::Hash, BlockStats> for Dev<Block, Client>
+where
+	Block: BlockT + 'static,
+	Client: BlockBackend<Block> + ProvideRuntimeApi<Block> + Send + Sync + 'static,
```
```suggestion
	Client: BlockBackend<Block> + HeaderBackend<Block> + ProvideRuntimeApi<Block> + Send + Sync + 'static,
```

bkchr commented on client/rpc/src/dev/mod.rs:
```diff
@@ -0,0 +1,116 @@
+				return Ok(None)
+			}
+		};
+		let parent_block = {
+			let parent_hash = *block.header().parent_hash();
+			let parent_block = self
+				.client
+				.block(&BlockId::Hash(parent_hash))
+				.map_err(|e| Error::BlockQueryError(Box::new(e)))?;
+			if let Some(parent_block) = parent_block {
+				parent_block.block
+			} else {
+				return Ok(None)
+			}
+		};
```
```suggestion
		let parent_header = {
			let parent_hash = *block.header().parent_hash();
			let parent_header = self
				.client
				.block(BlockId::Hash(parent_hash))
				.map_err(|e| Error::BlockQueryError(Box::new(e)))?;
			if let Some(header) = parent_header {
				header
			} else {
				return Ok(None)
			}
		};
```

bkchr commented on client/rpc/src/dev/mod.rs:
```diff
@@ -0,0 +1,116 @@
+		};
+		let block_len = block.encoded_size() as u64;
+		let num_extrinsics = block.extrinsics().len() as u64;
+		let pre_root = *parent_block.header().state_root();
```
```suggestion
		let pre_root = *parent_header.state_root();
```

bkchr commented on client/rpc/src/dev/mod.rs:
```diff
@@ -0,0 +1,116 @@
+		let mut runtime_api = self.client.runtime_api();
+		runtime_api.record_proof();
+		runtime_api
+			.execute_block(&BlockId::Hash(parent_block.hash()), block)
```
```suggestion
			.execute_block(&BlockId::Hash(parent_header.hash()), block)
```

bkchr commented on client/rpc/src/dev/mod.rs:
```diff
@@ -0,0 +1,116 @@
+		let witness_compact = witness
+			.into_compact_proof::<HasherOf<Block>>(pre_root)
+			.map_err(|_| Error::WitnessCompactionFailed)?
+			.encode();
```
```suggestion
			.encoded_size() as u64;
```

bkchr commented on client/rpc/src/dev/mod.rs:
```diff
@@ -0,0 +1,116 @@
+			.encode();
+		Ok(Some(BlockStats {
+			witness_len,
+			witness_compact_len: witness_compact.len() as u64,
```
```suggestion
			witness_compact_len,
```

bkchr commented on client/rpc-api/src/dev/error.rs:
```diff
@@ -0,0 +1,71 @@
+use crate::errors;
+use jsonrpc_core as rpc;
+
+/// Chain RPC Result type.
+pub type Result<T> = std::result::Result<T, Error>;
+
+/// Chain RPC future Result type.
+pub type FutureResult<T> = jsonrpc_core::BoxFuture<Result<T>>;
+
+/// Chain RPC errors.
```
```suggestion
/// Dev RPC Result type.
pub type Result<T> = std::result::Result<T, Error>;

/// Dev RPC future Result type.
pub type FutureResult<T> = jsonrpc_core::BoxFuture<Result<T>>;

/// Dev RPC errors.
```

bkchr commented on client/rpc-api/src/dev/error.rs:
```diff
@@ -0,0 +1,71 @@
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Error helpers for Chain RPC module.
```
```suggestion
//! Error helpers for Dev RPC module.
```

bkchr commented on client/rpc-api/src/dev/error.rs:
```diff
@@ -0,0 +1,71 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2017-2022 Parity Technologies (UK) Ltd.
```
```suggestion
// Copyright (C) 2022 Parity Technologies (UK) Ltd.
```

bkchr commented on client/rpc-api/src/dev/mod.rs:
```diff
@@ -0,0 +1,39 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2017-2022 Parity Technologies (UK) Ltd.
```
```suggestion
// Copyright (C) 2022 Parity Technologies (UK) Ltd.
```

bkchr commented on client/rpc-api/src/dev/mod.rs:
```diff
@@ -0,0 +1,39 @@
+/// This API contains unstable and unsafe methods only meant for development nodes. They
+/// are all flagged as unsafe for this reason.
+#[rpc]
+pub trait DevApi<Hash, BlockStats> {
```
Why are we generic over `BlockStats`? As said for in the runtime crate, we should move the definition of the type here above the trait.

bkchr commented on client/rpc-api/src/dev/mod.rs:
```diff
@@ -0,0 +1,39 @@
+pub trait DevApi<Hash, BlockStats> {
+	/// Reexecute the specified `block_hash` and gather statistics while doing so.
+	///
+	/// This function will require the specified block and its parent to be available
+	/// at the queried node. The latter is needed to generate the compact proof.
+	/// For more information see [`sp_runtime::BlockStats`].
```
```suggestion
	/// This function will require the specified block and its parent to be available
	/// at the queried node. If either the specified block or the parent are not available, 
	/// this function will return `None`.
```

The parent is also required for having its state to re-execute the block on this state.

bkchr reviewed (approved):
Most of the stuff are just nitpicks, otherwise it looks good.

dvdplm commented on client/rpc-api/src/dev/mod.rs:
```diff
@@ -0,0 +1,62 @@
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Substrate dev API.
```
```suggestion
//! Substrate dev API containing RPCs that are mainly meant for debugging and stats collection for developers. The endpoints in this RPC module are not meant to be available to non-local users and are all marked `unsafe`.
```

dvdplm commented on client/rpc-api/src/dev/mod.rs:
```diff
@@ -0,0 +1,62 @@
+pub struct BlockStats {
+	/// The length in bytes of the storage proof produced by executing the block.
+	pub witness_len: u64,
+	/// The length in bytes of the storage proof after it was compacted.
```
```suggestion
	/// The length in bytes of the storage proof after compaction.
```

(To me "after compression" sounds better, but then it should be `witness_compressed_len` too… Either is fine.)

athei replied:
We call it compact because there is compression, too. We compact the proof and then we compress it (with zstd). So using the word compression here would be confusing.

dvdplm commented on client/rpc-api/src/dev/mod.rs:
```diff
@@ -0,0 +1,62 @@
+	pub witness_compact_len: u64,
+	/// Length of the block in bytes.
+	///
+	/// This information could also be acquired by downloading the whole block. This is merely
+	/// a convenience to save some complexity on the client side.
```
```suggestion
	/// This information can also be acquired by downloading the whole block. This merely
	/// saves some complexity on the client side.
```

dvdplm commented on client/rpc-api/src/dev/mod.rs:
```diff
@@ -0,0 +1,62 @@
+	pub block_len: u64,
+	/// Number of extrinsics in the block.
+	///
+	/// This information could also be acquired by downloading the whole block. This is merely
+	/// a convenience to save some complexity on the client side.
```
```suggestion
	/// This information can also be acquired by downloading the whole block. This merely
	/// saves some complexity on the client side.
```

dvdplm commented on client/rpc-api/src/dev/mod.rs:
```diff
@@ -0,0 +1,62 @@
+pub trait DevApi<Hash> {
+	/// Reexecute the specified `block_hash` and gather statistics while doing so.
+	///
+	/// This function will require the specified block and its parent to be available
+	/// at the queried node. If either the specified block or the parent are not available,
+	/// this function will return `None`.
```
```suggestion
	/// This function requires the specified block and its parent to be available
	/// at the queried node. If either the specified block or the parent is pruned,
	/// this function will return `None`.
```

Is there any other reason than pruning that a block is unavailable?

dvdplm commented on client/rpc/src/dev/mod.rs:
```diff
@@ -0,0 +1,117 @@
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Substrate blockchain API.
```
```suggestion
//! Implementation of the [`DevApi`] trait providing debug utilities for Substrate based blockchains.
```

dvdplm reviewed (approved):
Code LGTM. Left some nitpicks. :)

athei commented:
bot merge
